flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -1487,7 +1487,7 @@ class JobManager(
           }
         }
 
-        eg.fail(cause)
+        eg.cancel()
```

Comment:
Yes, currently it is the case that state is removed from ZooKeeper when you shut down the cluster and run `postStop`. Do you think that this is a problem?

I've created https://issues.apache.org/jira/browse/FLINK-3411 last week for a related issue: if all execution retries (current default with checkpointing is `Integer.MAX_VALUE`) are exhausted, state is removed as well.